### PR TITLE
add support negative balance

### PR DESCRIPTION
**Pull request by dtsarenko**

### CODE REVIEW

File: src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java [hunk 1]
```diff
@@ -334,19 +339,42 @@ private BigDecimal chargingPockets(BigDecimal amountToBalanceCheckout, List<Pock
...
+        }
+
         return amountToBalanceCheckout;
     }
 
     private void assertNotEmpty(Balance balance, BigDecimal amount, BigDecimal amountToCheckout, List<Pocket> pockets) {
         if (pockets.isEmpty()) {
-            throw new NoEnoughMoneyException(balance.getId(), amount.subtract(amountToCheckout));
+            BalanceSpec.BalanceTypeSpec balanceTypeSpec = balanceSpecService.getBalanceSpec(balance.getTypeKey());
+            if (Boolean.FALSE.equals(balanceTypeSpec.isAllowNegative())) {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java]: тут несовсем верно, так как null не equals false

Comment (dtsarenko) [reply src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java]: boolean allowNegative = false;

Comment (dtsarenko) [reply src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java]: changed on  if (!balanceTypeSpec.isAllowNegative()) {}

File: src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java [hunk 2]
```diff
@@ -334,19 +340,42 @@ private BigDecimal chargingPockets(BigDecimal amountToBalanceCheckout, List<Pock
                 break;
             }
         }
+
+        if (amountToBalanceCheckout.compareTo(ZERO) > 0 && pockets.isLast() && isAllowNegative) {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java]: add negative pocket

**Review by sergeysenja1992 — APPROVED** [review]

**Review by VovaViter — APPROVED** [review]

File: src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java [hunk 3]
```diff
@@ -309,15 +314,22 @@ public Metadata mergeMetadata(Metadata metadata, Metadata additionalMetadata) {
...
         return pocketRepository.findPocketForChargingOrderByDates(balance, pageable);
     }
 
     private BigDecimal chargingPockets(BigDecimal amountToBalanceCheckout, List<PocketCharging> affectedPockets,
-                                       List<Pocket> pockets, BalanceChangeEvent changeEvent) {
+        Page<Pocket> pockets, BalanceChangeEvent changeEvent, AllowNegative allowNegative, Balance balance) {
 
         for (Pocket pocket : pockets) {
             BigDecimal pocketAmount = pocket.getAmount();
+            if (pocketAmount.compareTo(ZERO) <= 0) {
```
Comment (dtsarenko) [inline src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java:330]: @sergeysenja1992 pay attention, please

Comment (sergeysenja1992) [reply src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java:330]: maybe better to check pocket label?

**Pull request merged**